File: 105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return create(preorder, inorder, 0, preorder.size()-1, 0, inorder.size()-1);
    }
    
    TreeNode* create(vector<int>& preorder, vector<int>& inorder, int preStart, int preEnd, int inStart, int inEnd) {
        if (preStart > preEnd) return NULL;
        TreeNode* node = new TreeNode(preorder[preStart]);
        int i = inStart;
        for (; i <= inEnd; i++) {
            if (inorder[i] == node->val)
                break;
        }
        node->left = create(preorder, inorder, preStart + 1, preStart + i - inStart, inStart, i - 1);
        node->right = create(preorder, inorder, preEnd + i - inEnd + 1, preEnd, i + 1, inEnd);
        return node;
    }
};

File: 105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if inorder:
            ind =  inorder.index(preorder.pop(0))
            node = TreeNode(inorder[ind])
            node.left = self.buildTree(preorder, inorder[:ind])
            node.right = self.buildTree(preorder, inorder[ind + 1:])
            return node
        return None
